lib/http_clients/creditas/token_server.ex

```diff
@@ -0,0 +1,114 @@
+  First setup it on your `application.ex`, e.g.:
+
+  ```elixir
+  config = [
+    url: "https://auth-staging.creditas.com.br/api/internal_clients/tokens",
+    credentials: %{
+      username: "someuser",
+      password: "password123",
+      grant_type: "password"
+    }
+  ]
+  children = [{HttpClients.Creditas.TokenServer, name: MyApp.CreditasTokenServer, config: config}]
+  Supervisor.start_link(children, name: MyApp.Supervisor)
+  ```
+
+  Then you can retrieve the token:
+
+  ```elixir
+  # get token and refresh it if is expired or at most 30 seconds before expiring
+  HttpClients.Creditas.TokenServer.get_token(MyApp.CreditasTokenServer)
+
+  # get token and refresh it if is expired or at most 120 seconds before expiring
+  HttpClients.Creditas.TokenServer.get_token(MyApp.CreditasTokenServer, 120)
+
+  # get a new refreshed token
+  HttpClients.Creditas.TokenServer.get_new_token(MyApp.CreditasTokenServer)
+  ```
+  """
+
+  use Agent
+  require Logger
+
+  alias HttpClients.Creditas.Token
+
+  defguardp is_token_server(server) when is_pid(server) or is_atom(server)
+
+  @spec start_link(keyword()) :: {:ok, pid()} | {:error, any()} | no_return()
+  def start_link(opts) when is_list(opts) do
+    {config, opts} = Keyword.pop!(opts, :config)
+
+    Agent.start_link(
+      fn ->
+        {:ok, token} = request_new_token(config)
+        Logger.info("#{__MODULE__} started with pid #{inspect(self())}")
+        %{token: token, config: config}
+      end,
+      opts
+    )
+  end
+
+  defp request_new_token(url: url, credentials: credentials)
+       when is_binary(url) and is_map(credentials) do
+    case Tesla.post(creditas_client(), url, credentials) do
+      {:ok, %Tesla.Env{status: 201, body: token}} -> {:ok, build_token(token)}
+      {:ok, %Tesla.Env{} = response} -> {:error, response}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp creditas_client do
+    middleware = [
+      Tesla.Middleware.JSON,
+      {Tesla.Middleware.Retry, delay: 1_000, max_retries: 3},
+      {Tesla.Middleware.Timeout, timeout: 15_000}
+    ]
+
+    Tesla.client(middleware)
+  end
+
+  defp build_token(%{} = payload) do
+    now = DateTime.utc_now()
+    expires_at = DateTime.add(now, payload["expires_in"], :second)
+
+    %Token{
+      access_token: payload["access_token"],
+      expires_at: expires_at
+    }
+  end
+
+  @doc "Gets a new token from the given TokenServer"
+  @spec get_new_token(atom() | pid()) :: {:error, any} | {:ok, Token.t()}
+  def get_new_token(server) when is_token_server(server) do
+    config = Agent.get(server, & &1[:config])
+
+    with {:ok, token} <- request_new_token(config),
+         :ok <- Agent.update(server, &Map.put(&1, :token, token)) do
+      Logger.info("Creditas token updated for #{inspect(server)}")
+      {:ok, token}
+    end
+  end
+
+  @doc "Gets a token from the given TokenServer"
+  @spec get_token(atom() | pid(), integer()) :: map()
+  def get_token(server, seconds_before_refresh \\ 30)
```

Comment:
Conforme conversamos, será que n facilitaria o uso se tivermos só uma função de requisitar token?
Pra suportar o refresh talvez um parametro de refresh na `get_token`?
```suggestion
  @doc "Gets a token from the given TokenServer"
  @spec get_token(atom() | pid(), integer()) :: map()
  def get_token(server, seconds_before_refresh \\ 30, force_refresh \\ false)
```
@bcredi/backenders  oq acham?

Reply:
o que curto da separação atual é que o teste do `get_new_token` fica mais focado em recuperar/setar o token no Agent enquanto o `get_token` fica mais focado em testar os tempos ali de refresh.
mas indiferente pra mim, se acharem melhor ter uma função eu faço aqui

Reply:
feito do jeito sugerido, só deixei como uma Keyword pra dar mais a ideia de que são opts mesmo